### ISSUE TITLE
TAB key to indent/insert tab character

### ISSUE THREAD

**Issue by hongymagic** [OscarGodson/EpicEditor#159]
Hi,

I have hacked up initial version of `TAB` insertion in [my own fork](https://github.com/hongymagic/EpicEditor/commit/efdee33ec96e66968415658becfae3e2a890b5f0) (note: it's initial and does not support `shift-tab` combination yet).

I just wanted to start up a discussion around supporting `TAB` keys. The reason why I ask is because it has come to my attention that not everybody agrees with `TAB` key being used to indent; especially in the context of a webpage. _So what do people think about adding support for `TAB`?_

Ideally, in the context of an editor, I would use `TAB` key to indent selected lines or for (de)indentation. We can make use of [tab-size](https://developer.mozilla.org/en/CSS/-moz-tab-size) CSS property to adjust tab-width.

**Comment by OscarGodson** [comment]
I for one love this idea. :+1:

I wanted to get this in there eventually. I'm almost certain people will want to disable this tho for some reason or another. Why not just throw a `smartTab` option in there? `smartTab` could default to `true` and "smart" tab because it makes tabs work like a real text editor.

It'd also be super awesome if you could get it so it remembers the indentation of the last line. So, if the `|` is the cursor:

```
text...
    a tabbed line and I press enter the cursor ends up...
    |
```

Anyway, when you think you're ready send a pull! This is looking great. Curious, why are there a bunch of empty lines added/deleted?

**Comment by hongymagic** [comment]
As @OscarGodson said:

> Why not just throw a smartTab option in there? smartTab could default to true and "smart" tab because it makes tabs work like a real text editor.

That is one way to handle differences I guess! I will gladly add another option `smartTab`.

> It'd also be super awesome if you could get it so it remembers the indentation of the last line.

I agree this would good too. Will definitely take this into account.

> Curious, why are there a bunch of empty lines added/deleted?

I noticed that too, I am assuming my vim settings didn't match EOL of the file. Don't worry, I will create a new branch and make sure current code/file formatting rules are followed.

Thanks!

**Comment by OscarGodson** [comment]
No worries! Was just curious. Do you think it should be plural or singular? `smartTab` or `smartTabs`? I'm leaning towards plural now that I say it out loud.

**Comment by hongymagic** [comment]
> Do you think it should be plural or singular? smartTab or smartTabs?

`smartTabs` should do the trick =)

> Curious, why are there a bunch of empty lines added/deleted?

It turns out, those lines that were deleted weren't entirely empty :) I've set my vim to remove any trailing spaces on `save`. Disabling it seems to leave those lines as-is.

**Comment by OscarGodson** [comment]
Oh, yeah, I see them too in my vim (they're highlighted red) its cool to
strip them out. No worries.

**Comment by doginthehat** [comment]
+1 gimme tabs!

(it's the first thing I tried in the editor and was a tad disappointed it didn't work)

**Comment by OscarGodson** [comment]
@doginthehat Yep. I want it to. I write lots of code docs so it kinda sucks not having the feature. I've just never gotten to it because the normal end user probably wont ever use the feature since it'll mostly be used in CMSs. But, yes, @hongymagic magic seems to be on it ;)

**Comment by doginthehat** [comment]
Cool. Maybe @hongymagic could submit a pull request?

**Comment by OscarGodson** [comment]
Yeah that's what he's working towards. He has 3 feature branches already and committing code! :thumbsup:

**Comment by vjpr** [comment]
+1

**Comment by tonytonyjan** [comment]
+1

**Comment by hellerbarde** [comment]
What's the status on this? Can I help? I'd love to help with this whole Tab thing. (@hongymagic didnt have an issuetracker)

**Comment by OscarGodson** [comment]
@hellerbarde Sure, I'm not actively working on this problem and @hongymagic gave me a pull request that I'm still testing for something unrelated so, if you want, go ahead and send me a pull request!

**Comment by hellerbarde** [comment]
arright. i'll see what I can do. :)

**Comment by OscarGodson** [comment]
@hellerbarde :heart::heart::heart::heart::heart::heart::heart::heart::heart::heart::heart::heart::heart:

**Comment by OscarGodson** [comment]
@hellerbarde whatever happened with this?

**Comment by hellerbarde** [comment]
I started working on it, but I wasn't able to insert the spaces or tabs. I had a look at hongymagic's code, but I didn't really understand it. And then midterms started ;) 

I'm working on my Javascript-fu, but it's not where it needs to be yet.

**Comment by OscarGodson** [comment]
@hellerbarde well, thanks for trying :)

**Comment by hellerbarde** [comment]
I'm not giving up just yet :) 

But a few things will be very tricky. Ideally, it should be possible to select a bunch of lines and hit Tab to indent all of them. But because EpicEditor actually uses a content-editable, not every line is it's own html element. (which is totally fine, of course). I took a quick look at how Codemirror does it, and they have a completely different approach which is much more complex but gives them a lot of flexibility. I might try and find a half-way compromise without having to gut the current EpicEditor plumbings, because right now it's really nice, clean and elegant.

**Comment by valx76** [comment]
Please look at this script : https://github.com/julianlam/tabIndent.js

Doesn't require any library like jQuery and works great (with some cool features, more developper oriented) !
Just have to adapt it for editable body element..

**Comment by svenjacobs** [comment]
+1 for smart tabs

**Comment by MicrowaveDev** [comment]
I tried to do was not very pretty, but it works:
https://github.com/OscarGodson/EpicEditor/pull/386
